Add test cases for testing additional Presets write and commit constr…

nivi-apple opened the pull request:
…aints

- Add a test for adding a preset with a preset scenario not present in PresetTypes

- Add a test for testing addition of presets such that the total number of presets added is greater than the total number of presets supported

Fixes: #34695

nivi-apple committed:
Add test cases for testing additional Presets write and commit constraints

- Add a test for adding a preset with a preset scenario not present in PresetTypes

- Add a test for testing addition of presets such that the total number of presets added is greater than the total number of presets supported

ReneJosefsen commented on src/python_testing/TC_TSTAT_4_2.py:
```diff
@@ -461,7 +465,41 @@ async def test_TC_TSTAT_4_2(self):
+            # Send the AtomicRequest begin command
+            await self.send_atomic_request_begin_command()
+
+            await self.write_presets(endpoint=endpoint, presets=test_presets, expected_status=Status.ResourceExhausted)
```
Should we add a rollback here as well?

nivi-apple replied:
maybe as a separate PR. i will add a new issue and tests

nivi-apple replied:
done.

nivi-apple committed:
Modify the number of presets supported test case to read the number of presets supported and build a preset list whose size exceeds that to test

nivi-apple committed:
Modify the number of presets supported test case to read the number of presets supported and build a preset list whose size exceeds that to test

nivi-apple committed:
Add rollback after test step 18

bzbarsky-apple commented on src/python_testing/TC_TSTAT_4_2.py:
```diff
@@ -461,7 +465,44 @@ async def test_TC_TSTAT_4_2(self):
+        self.step("18")
+        if self.pics_guard(self.check_pics("TSTAT.S.F08") and self.check_pics("TSTAT.S.A0050") and self.check_pics("TSTAT.S.Cfe.Rsp")):
+
+            # Read the active preset handle attribute and verify it was updated to preset handle
```
That is ... not the attribute that is being read?

nivi-apple replied:
fixed.

bzbarsky-apple commented on src/python_testing/TC_TSTAT_4_2.py:
```diff
@@ -461,7 +465,44 @@ async def test_TC_TSTAT_4_2(self):
+            logger.info(f"Rx'd PresetTypes: {presetTypes}")
+
+            test_presets = copy.deepcopy(new_presets_with_handle)
+            test_presets.append(cluster.Structs.PresetStruct(presetHandle=NullValue, presetScenario=cluster.Enums.PresetScenarioEnum.kWake,
```
How do you know adding 4 presets here is enough to hit "greater than the number of presets supported"?

Why was PresetTypes read at all above?

Shouldn't we be reading something that tells us the number of presets supported, then building a list that is longer than that?

nivi-apple replied:
done.

nivi-apple committed:
Merge remote-tracking branch 'upstream/master' into add_test_scripts_and_plans_for_preset_checks

nivi-apple committed:
Update thermostat-delegate-impl.h

bzbarsky-apple commented on src/python_testing/TC_TSTAT_4_2.py:
```diff
@@ -461,8 +465,62 @@ async def test_TC_TSTAT_4_2(self):
+                            break
+                        else:
+                            didCopyPreset = False
+                    if not didCopyPreset:
```
This should probably be checking whether we actually have NumberOfPresets (from the PresetTypeStruct) presets of this type, and if not adding until we do.

nivi-apple replied:
done.

nivi-apple committed:
Address review comments

nivi-apple committed:
Add support to check for numberOfPresets supported for each preset type and build the presets list with multiple presets of each type

nivi-apple commented on src/python_testing/TC_TSTAT_4_2.py:
```diff
@@ -461,8 +465,62 @@ async def test_TC_TSTAT_4_2(self):
+        self.step("17")
+        if self.pics_guard(self.check_pics("TSTAT.S.F08") and self.check_pics("TSTAT.S.A0050") and self.check_pics("TSTAT.S.Cfe.Rsp")):
+            test_presets = new_presets_with_handle.copy()
+            test_presets.append(cluster.Structs.PresetStruct(presetHandle=NullValue, presetScenario=9,
```
fixed.

nivi-apple commented on src/python_testing/TC_TSTAT_4_2.py:
```diff
@@ -461,8 +465,62 @@ async def test_TC_TSTAT_4_2(self):
+            # Read the Presets to copy the existing presets into our testPresets list below.
+            presets = await self.read_single_attribute_check_success(endpoint=endpoint, cluster=cluster, attribute=cluster.Attributes.Presets)
 
+            # Since we currently support one preset of each type, run the test only if we have enough preset types to create a preset list whose length exceeds the
```
i modified the comment to say sdk instead of "we". Also there is already an issue to support multiple presets of each type in the sdk - https://github.com/project-chip/connectedhomeip/issues/34556. will fix this as part of that. added a comment in the issue to track this.

restyled-commits committed:
Restyled by autopep8

nivi-apple committed:
Fix log line formatting

nivi-apple commented on src/python_testing/TC_TSTAT_4_2.py:
```diff
@@ -461,7 +465,86 @@ async def test_TC_TSTAT_4_2(self):
+        self.step("17")
+        if self.pics_guard(self.check_pics("TSTAT.S.F08") and self.check_pics("TSTAT.S.A0050") and self.check_pics("TSTAT.S.Cfe.Rsp")):
+            test_presets = new_presets_with_handle.copy()
+            test_presets.append(cluster.Structs.PresetStruct(presetHandle=NullValue, presetScenario=cluster.Enums.PresetScenarioEnum.kGoingToSleep,
```
I see how this could be a problem. gotcha. i need to broaden my thinking

nivi-apple replied:
done.

nivi-apple committed:
Update src/python_testing/TC_TSTAT_4_2.py

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

nivi-apple committed:
Fix test step 17 to find a preset scenario in PresetScenarioEnum that is not present in PresetTypes to run the test

- Fix test step 18 to build a presets list that exceeds the number of presets supported correctly

bzbarsky-apple commented on src/python_testing/TC_TSTAT_4_2.py:
```diff
@@ -461,7 +465,65 @@ async def test_TC_TSTAT_4_2(self):
+                            ++presetsAddedForScenario
+
+                    while presetsAddedForScenario < presetType.numberOfPresets:
+                            testPresets.append(cluster.Structs.PresetStruct(presetHandle=NullValue, presetScenario=scenario,
+                                                                            name="Preset", coolingSetpoint=2500, heatingSetpoint=1700, builtIn=False))
+                            ++presetsAddedForScenario
```
```suggestion
                        testPresets.append(cluster.Structs.PresetStruct(presetHandle=NullValue, presetScenario=scenario,
                                                                        name="Preset", coolingSetpoint=2500, heatingSetpoint=1700, builtIn=False))
                        ++presetsAddedForScenario
```

nivi-apple replied:
just took the restyled suggestion to make it easier for me

hasty commented on src/python_testing/TC_TSTAT_4_2.py:
```diff
@@ -461,7 +465,65 @@ async def test_TC_TSTAT_4_2(self):
+                    for preset in presets:
+                        if scenario == preset.presetScenario:
+                            testPresets.append(preset)
+                            ++presetsAddedForScenario
```
Does this actually work? I didn't think Python had increment operators, as ints are immutable. Or do we have some syntactic sugar library floating around somewhere that's making this go?

nivi-apple replied:
no it doesn't. i fixed it

nivi-apple replied:
sorry noob at python

restyled-commits committed:
Restyled by autopep8

nivi-apple committed:
Fix lint errors

bzbarsky-apple commented on src/python_testing/TC_TSTAT_4_2.py:
```diff
@@ -461,7 +465,86 @@ async def test_TC_TSTAT_4_2(self):
+                            testPresets.append(preset)
+                            presetsAddedForScenario = presetsAddedForScenario + 1
+
+                    if presetsAddedForScenario < presetType.numberOfPresets:
```
```suggestion
                    while presetsAddedForScenario < presetType.numberOfPresets:
```
Otherwise this does not in fact work right.

nivi-apple replied:
yeah since number of presets is 1 it passed. yikes fixed it

nivi-apple committed:
Add a while loop to add more presets until max is reached
